### ISSUE TITLE
surprising treatment of unspecified cflags

### ISSUE THREAD

**Issue by mwkrentel** [LLNL/spack#6096]
I think spack's handling of default (unspecified) CFLAGS and CXXFLAGS
is surprising and incorrect.

Let's use xz utils (lzma compression) as a simple example.  xz is
straightforward autotools and has a tiny package.py script, using all
default AutotoolsPackage build machinery.

First, let's build xz manually (outside of spack), specifying only
--prefix and not CFLAGS.

$ ../xz-5.2.3/configure  --prefix=/home/krentel/work/install/
$ make

The generated Makefile has:

CFLAGS = -g -O2

And a typical compile line is (note the inclusion of -g -O2).

libtool: compile: gcc -std=gnu99 -DHAVE_CONFIG_H
-I. -I../../../xz-5.2.3/src/liblzma
-I../.. -I../../../xz-5.2.3/src/liblzma/api
-I../../../xz-5.2.3/src/liblzma/common
-I../../../xz-5.2.3/src/liblzma/check
-I../../../xz-5.2.3/src/liblzma/lz
-I../../../xz-5.2.3/src/liblzma/rangecoder
-I../../../xz-5.2.3/src/liblzma/lzma
-I../../../xz-5.2.3/src/liblzma/delta
-I../../../xz-5.2.3/src/liblzma/simple -I../../../xz-5.2.3/src/common
-DTUKLIB_SYMBOL_PREFIX=lzma_ -pthread -fvisibility=hidden -Wall
-Wextra -Wvla -Wformat=2 -Winit-self -Wmissing-include-dirs
-Wstrict-aliasing -Wfloat-equal -Wundef -Wshadow -Wpointer-arith
-Wbad-function-cast -Wwrite-strings -Wlogical-op -Waggregate-return
-Wstrict-prototypes -Wold-style-definition -Wmissing-prototypes
-Wmissing-declarations -Wmissing-noreturn -Wredundant-decls -g -O2 -MT
liblzma_la-common.lo -MD -MP -MF .deps/liblzma_la-common.Tpo -c
../../../xz-5.2.3/src/liblzma/common/common.c -fPIC -DPIC -o
.libs/liblzma_la-common.o

So, I didn't specify CFLAGS and xz's configure script selected
CFLAGS='-g -O2' as its default.

But now, if we try spack install without specifying cflags, we get a
very different result.

$ spack  install  -v  --keep-stage  xz

Using --keep-stage, we go into the build directory and look at the
generated Makefile.

CFLAGS = 

And indeed, a compile line makes no mention of -g or -O2.

libtool: compile: /home/krentel/work/spack-repo/lib/spack/env/gcc/gcc
-std=gnu99 -DHAVE_CONFIG_H
-I. -I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/liblzma
-I../.. -I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/liblzma/api
-I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/liblzma/common
-I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/liblzma/check
-I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/liblzma/lz
-I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/liblzma/rangecoder
-I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/liblzma/lzma
-I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/liblzma/delta
-I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/liblzma/simple
-I/home/krentel/work/spack-repo/var/spack/stage/xz-5.2.3-zaiqz5siodty64rxu4mqz4ayjogfbabw/xz-5.2.3/src/common
-DTUKLIB_SYMBOL_PREFIX=lzma_ -pthread -fvisibility=hidden -Wall
-Wextra -Wvla -Wformat=2 -Winit-self -Wmissing-include-dirs
-Wstrict-aliasing -Wfloat-equal -Wundef -Wshadow -Wpointer-arith
-Wbad-function-cast -Wwrite-strings -Wlogical-op -Waggregate-return
-Wstrict-prototypes -Wold-style-definition -Wmissing-prototypes
-Wmissing-declarations -Wmissing-noreturn -Wredundant-decls -MT
liblzma_la-common.lo -MD -MP -MF .deps/liblzma_la-common.Tpo -c
common/common.c -fPIC -DPIC -o .libs/liblzma_la-common.o

That is, the spack default is -O0 and no debugging symbols.

The problem is that spack is inserting an explicit CFLAGS='' (empty)
into the configure script.  Again, you can see this with --keep-stage
in config.log.  (Currently this is via the environment, not the
command line, but either way works the same.)

I think this is very surprising and actually incorrect.  Instead,
spack should recoginize that cflags is unspecified and pass along
'unspecified' by leaving off CFLAGS from the command line or the
environment.

If I put cflags='...' in the spec, then of course, spack should pass
that on to each package.  But if I omit cflags, then spack should omit
CFLAGS so that each package chooses its natural default.

I'm curious how spack users handle this.

Is everyone careful to always add cflags='-g -O2' to the spec?

Are people getting lots of unoptimized builds without debug symbols
and not knowing it?

Thanks,

--Mark

**Comment by michaelkuhn** [comment]
> Are people getting lots of unoptimized builds without debug symbols and not knowing it?

I would guess this is how it is. I have been using Spack for quite some time and never really noticed.

On top of that, Spack's behavior seems to be pretty inconsistent. As far as I can tell, this behavior is due to the `default_flag_handler` setting the environment variables. Currently, only autotools and cmake packages seem to do that. lua is a normal package and does not get its flags overridden.

On the one hand, I guess overriding the flags makes sense from a reproducibility point of view. On the other hand, this would be pretty easy to fix by changing the `default_flag_handler` to only set the flags if they are not empty.

**Comment by mwkrentel** [comment]
Yes, the problem as I described primarily affects autotools packages,
cmake is different.  But there are many, many autotools packages.

But that brings up an interesting question: if I don't specify a cmake
build type (Release, RelWithDebInfo, etc), does spack impose one on
me, or does it leave it unspecified for the package to choose its own
default.

**Comment by svenevs** [comment]
> if I don't specify a cmake build type (Release, RelWithDebInfo, etc), does spack impose one on me

Yes.  Currently it is `RelWithDebInfo`.  LLVM is an example of a package that defaults to release, though, as the debugging symbols typically create a >20GB installation.  See #5261, the introduction of default `RelWithDebInfo` squeaked that in x0  There is also a discussion in #5265 as to whether the default should just be `Release`.

> Is everyone careful to always add cflags='-g -O2' to the spec?

No.  I doubt anybody has really paid much attention :wink:  Moreover, the debugging symbols are in fact not useful unless you still have the source code around for a given installation.  Spack by default will not do this, so while your binary may have debugging symbols when you use `-g`, it's actually not very useful because the symbols point to the file where it was built (which was deleted).  Unless you re-stage the package or never delete any of the staging directories, the debugging symbols are just bloat.  Or at least that was my ultimate takeway from the linked issue over default to Release.

> The problem is that spack is inserting an explicit CFLAGS='' (empty) into the configure script.

That may be an oversight, or it may have an actual purpose.  The fix could be as simple as checking for the empty string and only adding it if needbe?

**Comment by pramodskumbhar** [comment]
> I actually think this is deliberate, to specifically override any CFLAGS environment variables a user may have set (clean build environment).

In order to have clean build we typically `unset` env variables in [build_environment.py](https://github.com/spack/spack/blob/develop/lib/spack/spack/build_environment.py). But not all env variables are unset there.

**Comment by mwkrentel** [comment]
Returning to my original question about the default treatment of
CFLAGS for autotools packages ....

The default CFLAGS for spack is empty string, which essentially means
no optimization, ie, -O0.  (At least, that's the GNU answer.  IIRC,
some vendor compilers will optimize by default.)  My questions are:

1. Is this really the intended semantics?

2. Is this a good idea?

I suggest -O0 is not a good idea and that -O or -O2 would be a better
choice.  The choice of -g depends on the use case.  Leaving it off by
default is reasonable enough, as long as I can turn it back on when I
want.

Either way, I suggest adding a more explicit note in the documentation
that adding cflags=... to the spec is a good idea, or else you'll
likely end up with something that you don't expect.

And note: I understand wanting to unset CFLAGS, etc, from the
environment.  But unsetting CFLAGS and explicitly setting it to empty
are very different.  If spack simply unset CFLAGS, then the choice
would fall back to the package default.  Explicitly setting it to
empty changes -O2 to -O0, and that's what I'm bringing up here.  I
consider that very surprising and actually incorrect.

**Comment by michaelkuhn** [comment]
I also agree that something like `-O2` would be a more sensible default. For example, Fedora uses `-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches` for its packages (plus a bit more when hardening is not disabled).

**Comment by mwkrentel** [comment]
I don't know that this is really addressed across all packages, and I
don't know that spack has the machinery to set optimization defaults
for each package type (autotools, cmake, etc).  (I'm concerned that
many packages still build at -O0.)  But the treatment of cflags has
been reworked to at least make it possible to set reasonable
optimization defaults on a per-package basis.

So, I'll close this and address things on a per-package basis.  But we
may want to revisit this issue at some point with better machinery to
set optimization defaults.

**Comment by mwkrentel** [comment]
See Greg Becker's commit 28d8784ab9 from Dec 2017 that added
flag_handler, inject_flags, env_flags and build_system_flags.  This
provides sufficient tools to handle C/C++ flags cleanly, at least on a
package-by-package basis.

I just retried the xz package, and it now builds with '-g -O2' by
default.  So maybe now most autotools packages use a sensible default.
I've really only checked 1 or 2 packages.

If I notice any widespread or systemic strangeness in the future, I'll
be sure to bring this up again.

And yes, cmake is different.

**Comment by svenevs** [comment]
Thank you for sharing @mwkrentel !
